File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 4/FestivalMuzicaJavaFX/src/main/java/ro/mpp2026/festivalmuzicajavafx/service/ShowsService.java
package ro.mpp2026.festivalmuzicajavafx.service;

import lombok.Getter;
import ro.mpp2026.festivalmuzicajavafx.domain.Show;
import ro.mpp2026.festivalmuzicajavafx.domain.ShowFilter;
import ro.mpp2026.festivalmuzicajavafx.domain.Ticket;
import ro.mpp2026.festivalmuzicajavafx.repository.ShowsDBRepository;
import ro.mpp2026.festivalmuzicajavafx.repository.ShowsRepository;
import ro.mpp2026.festivalmuzicajavafx.repository.TicketsDBRepository;
import ro.mpp2026.festivalmuzicajavafx.repository.TicketsRepository;
import ro.mpp2026.festivalmuzicajavafx.utils.JdbcUtils;
import ro.mpp2026.festivalmuzicajavafx.utils.Observable;
import ro.mpp2026.festivalmuzicajavafx.utils.Observer;

import java.io.IOException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;
import java.util.stream.StreamSupport;

public class ShowsService implements Observable {
    @Getter
    private static final ShowsService instance = new ShowsService();

    private final ShowsRepository showsRepository;
    private final TicketsRepository ticketsRepository;

    List<Observer> subscribers;

    private ShowsService() {
        Properties props = new Properties();
        try {
            var inStream = AuthService.class.getResourceAsStream("/ro/mpp2026/festivalmuzicajavafx/db.properties");
            if (inStream == null) {
                throw new RuntimeException("Cannot find db.properties in classpath at /ro/mpp2026/festivalmuzicajavafx/db.properties");
            }
            props.load(inStream);
        } catch (IOException e) {
            throw new RuntimeException("Error loading db.properties", e);
        }

        JdbcUtils jdbcUtils = new JdbcUtils(props);
        showsRepository = new ShowsDBRepository(jdbcUtils);
        ticketsRepository = new TicketsDBRepository(jdbcUtils);

        subscribers = new ArrayList<>();
    }

    public void save(String artistName, LocalDate date, String location, int remainingSeats) {
        Show show = new Show(0L, artistName, date, location, remainingSeats);
        showsRepository.save(show);
    }

    public List<Show> findAll() {
        return StreamSupport.stream(showsRepository.findAll().spliterator(), false)
                .toList();
    }

    public List<Show> findAll(ShowFilter showFilter) {
        return StreamSupport.stream(showsRepository.findAll(showFilter).spliterator(), false)
                .toList();
    }

    public Optional<Show> findById(Long id) {
        Show show = showsRepository.findOne(id);
        return Optional.ofNullable(show);
    }

    public void bookTicketForShow(Long showId, String client, Integer numberOfSeats) {
        Show show = showsRepository.findOne(showId);

        if(show.getRemainingSeats() < numberOfSeats)
            throw new RuntimeException("Not enough seats");

        Ticket ticket = new Ticket(0L, client, show, numberOfSeats);
        show.setRemainingSeats(show.getRemainingSeats() - numberOfSeats);

        ticketsRepository.save(ticket);
        showsRepository.update(showId, show);

        notifyObservers();
    }

    public void updateTicket(Long ticketId, String client, Long showId, Integer numberOfSeats) {
        Ticket ticket = ticketsRepository.findOne(ticketId);
        Show show = ticket.getShow();

        if(show.getRemainingSeats() + ticket.getNoSeats() - numberOfSeats < 0)
            throw new RuntimeException("Not enough seats");

        show.setRemainingSeats(show.getRemainingSeats() + ticket.getNoSeats() - numberOfSeats);
        ticket.setNoSeats(numberOfSeats);

        ticketsRepository.update(ticketId, ticket);
        showsRepository.update(show.getId(), show);

        notifyObservers();
    }

    public List<Ticket> findAllTickets() {
        return StreamSupport.stream(ticketsRepository.findAll().spliterator(), false)
                .toList();
    }

    public List<Ticket> findAllTicketsForShow(Long showId) {
        return StreamSupport.stream(ticketsRepository.findAllForShow(showId).spliterator(), false)
                .toList();
    }

    public int getNumberOfSoldSeatsForShow(Long showId) {
        return findAllTicketsForShow(showId).stream()
                .mapToInt(Ticket::getNoSeats)
                .sum();
    }

    @Override
    public void subscribe(Observer observer) {
        subscribers.add(observer);
    }

    @Override
    public void notifyObservers() {
        for(var observer : subscribers)
            observer.update();
    }
}

File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 5/FestivalMuzicaCSharpWFA/FestivalMuzicaCSharpWFA/Service/ShowsService.cs
using System.Configuration;
using FestivalMuzicaCSharp.Domain;
using FestivalMuzicaCSharp.Repository;
using FestivalMuzicaCSharpWFA.Utils;

namespace FestivalMuzicaCSharpWFA.Service;

public class ShowsService : IObservable {
    private static readonly ShowsService _instance = new  ShowsService();
    public static ShowsService Instance => _instance;
    private readonly IShowsRepository _showsRepository;
    private readonly ITicketsRepository _ticketsRepository;
    private List<IObserver> _observers;
    
    private ShowsService()
    {
        var props = new Dictionary<string, string>();
        var connStr = ConfigurationManager.ConnectionStrings["musicFestivalDB"]?.ConnectionString;
        if (connStr == null)
            throw new Exception("Cannot find connection string 'musicFestivalDB' in app.config");
        props["ConnectionString"] = connStr;
        
        _showsRepository = new ShowsDbRepository(props);
        _ticketsRepository = new TicketsDbRepository(props);
        
        _observers = new List<IObserver>();
    }

    public void Save(String artistName, DateOnly date, string location, int noSeats)
    {
        var show = new Show(0, artistName, date, location, noSeats);
        _showsRepository.Save(show);
        NotifyObservers();
    }
    
    public List<Show> FindAll()
    {
        return _showsRepository.FindAll().ToList();
    }

    public List<Show> FindAll(ShowFilter showFilter)
    {
        List<Show> shows = _showsRepository.FindAll().ToList();
        
        if(showFilter.ArtistName != null)
            shows = shows.Where(s => s.ArtistName.Contains(showFilter.ArtistName)).ToList();
        
        if(showFilter.Date != null)
            shows = shows.Where(s => s.Date.Equals(showFilter.Date)).ToList();
        
        if(showFilter.Location != null)
            shows = shows.Where(s => s.Location.Contains(showFilter.Location)).ToList();
        
        if(showFilter.RemainingSeats != null)
            shows = shows.Where(s => s.RemainingSeats.Equals(showFilter.RemainingSeats)).ToList();

        return shows;
    }
    
    public Show? FindOne(long id)
    {
        return _showsRepository.FindOne(id);
    }
    
    public void BookTicketForShow(long showId, string clientName, int noSeats)
    {
        var show = _showsRepository.FindOne(showId);
        if (show == null)
            throw new Exception("Show not found");
        
        if (show.RemainingSeats < noSeats)
            throw new Exception("Not enough seats available");

        var ticket = new Ticket(0, clientName, show, noSeats);
        show.RemainingSeats -= noSeats;

        _ticketsRepository.Save(ticket);
        _showsRepository.Update(showId, show);

        NotifyObservers();
    }
    
    public void UpdateTicket(long ticketId, string clientName, int noSeats)
    {
        var ticket = _ticketsRepository.FindOne(ticketId);
        if (ticket == null)
            throw new Exception("Ticket not found");
        
        var show = ticket.Show;
        int seatDifference = noSeats - ticket.NoSeats;

        if (show.RemainingSeats < seatDifference)
            throw new Exception("Not enough seats available");

        show.RemainingSeats -= seatDifference;
        ticket.ClientName = clientName;
        ticket.NoSeats = noSeats;

        _ticketsRepository.Update(ticketId, ticket);
        _showsRepository.Update(show.Id, show);

        NotifyObservers();
    }

    public List<Ticket> FindAllTickets()
    {
        return _ticketsRepository.FindAll().ToList();
    }
    
    public List<Ticket> FindAllTicketsForShow(long showId)
    {
        return _ticketsRepository.FindAll().Where(t => t.Show.Id == showId).ToList();
    }
    
    public int GetNumberOfSoldSeatsForShow(long showId)
    {
        return _ticketsRepository.FindAll().Where(t => t.Show.Id == showId).Sum(t => t.NoSeats);
    }
    
    public void Subscribe(IObserver observer)
    {
        if (!_observers.Contains(observer))
            _observers.Add(observer);
    }

    public void Unsubscribe(IObserver observer)
    {
        if (!_observers.Contains(observer))
            _observers.Remove(observer);
    }

    public void NotifyObservers()
    {
        foreach (var observer in _observers)
        {
            observer.Update();
        }
    }
}